docs: rewrite (#2521)
tigerbeetle/tigerbeetle: docs/internals/ARCHITECTURE.md

Reviewer: :exploding_head: This is a great write-up!

- are very contended and parallelize/shard poorly,
- consist mostly of writes,
- need very high throughput and moderately low latency,
- require strong consistency guarantees,
- and demand very high levels of durability.

At the moment, TigerBeetle is focused on financial transactions (transfers). Each transfer touches
two accounts, and accounts are Pareto-distributed --- a few accounts are responsible for the bulk of
the transfers. Multi-object transactions in combination with contention make traditional database
architecture inefficient --- locking the two accounts and then going over the network to the
application to compute the balances' update is slow, and, due to contention, this work can't be
efficiently distributed across separate machines.

While at the moment TigerBeetle implements only accounting logic, the underlying state machine can
be swapped for a different one. TigerBeetle does only double-entry bookkeeping, but _can_ do
anything that has the same combination of high performance and safety requirements, and high
contention.

## Overview

TigerBeetle is a distributed system. It is a cluster of six replicas. The state of each replica is

> just a single file on disk, called data file. Replicas use consensus protocol to ensure that their

Reviewer: ```suggestion
just a single file on disk, called the data file. Replicas use a consensus protocol to ensure that their
```

- converts the next request to a prepare,
- appends the prepare to the write ahead log (WAL), assigning it the next sequence number and
  adding a checksum "pointer" to the previous log entry,
- replicates prepare across backups.

When a backup receives a prepare, it writes it to the WAL section of the data file and sends a
`prepare_ok` message to the primary. Once the primary receives a quorum of `prepare_ok` messages,
the prepare is considered committed, meaning that it can no longer be removed from or reordered in
the log.

Replicas execute committed prepares in sequence number order, by applying a batch of transfers to
local state. Because all replicas start with the same (empty) state, and because the state
transition function is deterministic, the replicas arrive at the same state.

If a primary fails, the consensus algorithm proper ensures that a different replica becomes a
primary, and that the new primary correctly reconstructs the latest state of the log.

The derived state of the system is the append-only log of immutable transfers (past transfers are
stored for idempotence) and the current balances of all accounts. Physically on disk, the state is

> stored as a collection (a forest) of LSM trees. Each LSM tree is a sorted  collection of objects.

Reviewer: ```suggestion
stored as a collection (a forest) of LSM trees. Each LSM tree is a sorted collection of objects.
```

This approach gives good performance if most transactions are independent (e.g. are mostly reads).
For financial accounting, the opposite is true --- most transactions are writes, and they conflict
due to popular accounts. For this reason, TigerBeetle always executes a transaction directly inside
the database, avoiding moving data to code over the network.

### Single Thread

TigerBeetle is single-threaded. There are both positive and negative reasons for this design choice.

The primarily negative reason is that the underlying workload is inherently contentious: some
accounts are just way more popular than owners, and transfers between hot accounts inherently
sequentialize the system. Trying to make transactions parallel not only doesn't make it faster, in
fact, the overhead of synchronization tends to dominate useful work. Formally, our claim is that
financial transaction processing is an infinite-COST problem, using the terminology from [Frank
McSherry's paper](https://www.usenix.org/system/files/conference/hotos15/hotos15-paper-mcsherry.pdf).

The positive reason for using a single thread is that CPUs are quite fast actually, and the reports

> of Moore's Law demise are somewhat overstated. A single core can easily to 1mil TPS if:

Reviewer: ```suggestion
of Moore's Law demise are somewhat overstated. A single core can easily scale to 1mil TPS if:
```

constructor, and instead explicitly pass allocator to the specific methods that require allocation,
which is a perfect fit for the memory management strategy used in TigerBeetle. Cross compilation
that works and direct (glibc-less) bindings to the kernel help keep dependency count down.

Most importantly, this all is possible using very frugal language machinery. Zig lends itself very
well to low-abstraction first-order code that does the work directly, which makes it easy to author
and debug performance-oriented code.

### Determinism

A meta principle above "static allocation" is determinism --- a property that, given the same
inputs, the software not only gives the same logical result, but also arrives at the result using
the same physical path. There's no single specific reason to demand determinism, but it tends to
consistently simplify and improve the system. In general everything in TigerBeetle is deterministic,
but here are some of the places where that determinism leads to big advantages:

- Determinism greatly simplifies the implementation of a replicated state machine, as it reduces the

>   problem of synchronizing mutable state to a much simpler problem of synchronizing immutable,

Reviewer: ```suggestion
  problem of synchronizing mutable state to a much simpler problem of synchronizing an immutable,
```

  just a single disk block containing the problematic byte. This is the approach taken by
  TigerBeetle. It is guaranteed that replicas in the cluster converge on byte-for-byte identical LSM
  tree structure.
- Physical repair in turn massively simplifies error handing. The function that reads data from
  storage doesn't have an error condition. It _always_ receives the block with the requested
  checksum, but the block could be transparently read from a different replica.
- Physical determinism requires that LSM compaction work is scheduled deterministically, which
  ensures that compaction work is evenly spread throughout the operation, bounding the worst-case
  latencies.
- Determinism supercharges randomized testing, as any test failure can be reliably reproduced by
  sharing a seed that lead to the failure.

### Simulation Testing

TigerBeetle uses a variety of techniques to ensure that the code is correct, from good old
example-based tests to strict style guide which emphasizes simplicity and directness. The most
important technique deployed is simulation testing, as seen on <https://sim.tigerbeetle.com>.

> TigerBeetle simulator, VOPR, can run an entire cluster on a single thread, injecting various storage

Reviewer: ```suggestion
TigerBeetle's simulator, the VOPR, can run an entire cluster on a single thread, injecting various storage
```

help if you want to check if your code implements the algorithm correctly, or to verify that the
_assumptions_ behind the algorithm hold.

### Mechanical Sympathy

Mechanical Sympathy is the idea that, although a CPU is a general purpose device and can execute
anything, it is often possible to re-formulate a particular algorithm in a CPU-friendly manner,
which makes it much faster to run. Small details matter a lot for speed, and sometimes these small
details guide larger architecture.

In the context of TigerBeetle, mechanical sympathy spans all for primary colors of computation:

- Network
- Storage
- Memory
- CPU

Some examples:

Network has limited bandwidth. This means that if one node in the network requires much higher
bandwidth then the rest, the network will be underutilized. For this reason, when the primary needs
to replicate a prepare across the cluster, a ring topology is used: the primary doesn't broadcast
the prepare to every other replica, and instead sends it to just one other replica, relying on that
replica to forward the prepare farther.

Storage is typically capable of sustaining several parallel IO operations, so TigerBeetle tries to

> keep it utilized. For example, when compaction needs to read a table from disk, it enqueues writes

Reviewer: ```suggestion
keep it saturated. For example, when compaction needs to read a table from disk, it enqueues writes
```

(Or "busy".)

for several of the table's blocks at a time. At the same time, care is taken to not oversubscribe
the storage --- there's a limit on the maximum number of concurrent IO operations.

Memory is not fast --- operations that miss the cache and hit the memory are dramatically slower.
TigerBeetle data structures are compact, to maximize the chance that the data is cache, and are
organized around cache lines, such that the all data for a particular operation can be found in a
single cache line, as opposed to be more spread out in memory. For example, the size of a Transfer
object is two cache lines.

CPU can both sprint and parkour, but it is so much better at sprinting! CPU is very fast at straight

> line code, and can sweep several lanes at once with SIMD, but `if`s (especially unpredictable ones)

Reviewer: :laughing:

make it stagger. When processing events, where each event is either a new account or a new transfer,
TigerBeetle lifts the branching up. Events come in batches, and each batch is homogeneous ---
either all transfers, or all accounts. So the event-kind branching is moved out of the inner loop.

### Batching

One of the most effective optimization principles is the idea of amortizing the overhead --- if you
have to do a costly operation, make sure that its results "pay for" many smaller useful operation.
TigerBeetle uses batching at many different levels:

- Individual transfers are aggregated into a prepare, to amortize replication and prefetch IO
  overhead.
- Changes from 32 prepares are aggregated in memory before being written to disk together.
- Changes from 1024 prepares are aggregated before checkpoint is atomically advanced by overwriting
  the superblock.
- The LSM tree operates mostly in terms of tables and value blocks, each aggregating many individual
  records.

### LSM

TigerBeetle organizes data on disk as a collection of

> [Log Structured Merge Trees](https://www.youtube.com/watch?v=hkMkBZn2mGs). In context of

Reviewer: ```suggestion
[Log Structured Merge Trees](https://www.youtube.com/watch?v=hkMkBZn2mGs). In the context of
```

Batching is a special case of the more general principle of separating control plane and data plane.
Imagine yourself as a pilot in a cabin of a powerful jet. The cabin has all sorts of dials, levers
and buttons. By pressing the buttons, you control the jet engines. Most of the work of moving the
plane is done by the engines, the data plane. The pilot doesn't do the heavy lifting, they direct
the power of the engine.

In the context of TigerBeetle, deciding which prepare to apply would be control plane, but actually
going through each individual transfer is data plane. In general, control plane is O(1) to data
plane's O(N).

It is important to clearly separate the two mods of operation. Keeping control plane `if`s outside
of data plane `for`s keeps the inner loop free of branching, improving performance. Conversely, as
the overhead of control plane is small, it can use very aggressive assertions, up to spending O(N)
time to verify O(1) operation.

### Synchronous Execution

StateMachine's `commit` function, the one that actually implements double-entry accounting, is
synchronous. It takes a slice of transfers as input, and, in a single tight CPU loop, decides, for

> each transfers, whether it is applicable, computes the desired balance change if it is, and records

Reviewer: ```suggestion
each transfer, whether it is applicable, computes the desired balance change if it is, and records
```
